remove not-visible fields from forms

larkin opened the pull request:
This is a fix for #686:  it prevents accidental data loss that can be caused when form fields are not included in form_rules.

Each form rule needs to define a property for `visible_fields` which returns all fields that the rule exposes.  I've covered all the basic form rules.

@mrjoes how does this look?  I'm open to any / all feedback.

mrjoes commented on flask_admin/model/base.py:
```diff
@@ -1477,6 +1477,10 @@ def create_view(self):
             return redirect(return_url)
 
         form = self.create_form()
+        if self._form_create_rules:
```
I'm thinking if it should be really dynamic. Can do as a part of the form scaffolding. If someone will want dynamic form - they can always do it later.

Also, please move this logic into separate method, so it is callable and overridable.

larkin replied:
How a form comes to be: `self.edit_form` -> `self._edit_form_class` -> `self._refresh_forms_cache` -> `self.get_edit_form` -> `self.get_form` -> `self.scaffold_form`.

Where would you say scaffolding stops and dynamic begins-- in `_refresh_forms_cache`?

The problem I see is every point of scaffolding code beyond `_refresh_forms_cache` is implemented or overridden by code outside the base class; so it would require changing all the contrib modules to implement it-- seems bad for compatibility.

Without doing this dynamically, if anyone has existing form overrides, they're still susceptible to this bug, and we have no room to warn them.

In order to maintain separation between form generation and form rendering rules, I think doing this dynamically is best: it allows us to prevent data loss due to this bug no matter how someone implements form generation.  How you set up your form rendering rules should not cause you to lose data.

mrjoes replied:
It is slightly different. In method that calls `edit_form` and getting form class, it can call something like `self._validate_form_class` (or whatever name could be). `_validate_form_class` will call `self._check_render_rules` twice with default implementation that'll filter invisible fields out of the forms. 

Here's rationale:
1. Form can be supplied by the developer, without using Flask-Admin scaffolding. It might contain custom fields that contribute to the logic of the application without changing model itself (you can do all kind of stuff in the `on_model_change`, as form object is passed there). These fields should not be filtered out automatically.

2. Keeping previous point in mind, it should be easy to change default behavior. I'm not yet convinced that straight form class filtering is right default behavior. But agree that at least it should issue a warning. 

3. If developer wants to filter form using list of visible fields - he should be able to do it from any part of the code. In lots of cases there's no need to filter form in views - it won't change in most of the cases. If it is necessary to filter form dynamically, it is always possible to do it in `get_form` or create/edit overloads.

Hope it all makes sense.

larkin commented:
@mrjoes 

I added a check to `_refresh_cache` that validates each form class and sets a flag to show that the specific form has been validated already.  Then the create/edit view validates the form only if it hasn't been validated already.

In the worst case, this is dynamic validation and in the best case this is validation during scaffolding.  It still allows flexibility in how for form scaffolding overrides are applied (`edit_form`, `get_edit_form`, `scaffold_form`, etc.) without allowing accidental data loss.

If a developer supplies a form via their own means (e.g. is not using Flask-Admin scaffolding), and is also using form rules, a warning when a field is not included in the form rules is helpful in answering the question, "why can't I see this new field I added?".

What do you think?
